internal/compiler/program.go

```diff
@@ -865,6 +867,11 @@ func (p *Program) getOptionsDiagnosticsOfConfigFile() []*ast.Diagnostic {
 }
 
 func (p *Program) getSyntacticDiagnosticsForFile(ctx context.Context, sourceFile *ast.SourceFile) []*ast.Diagnostic {
+	// For JavaScript files, we report semantic errors for using TypeScript-only
+	// constructs from within a JavaScript file as syntactic errors.
+	if ast.IsSourceFileJS(sourceFile) {
+		return slices.Concat(p.getJSSyntacticDiagnosticsForFile(sourceFile), sourceFile.Diagnostics())
+	}
```

Comment:
Okay, there's a problem with this method. The JS diagnostic code races with binding as both touch the node flags at the same time.

The JS diagnostics code needs to move to the parser package (just move the file and fix its package clause), then the scan needs to be performed at the end of `finishSourceFile` for JS files, storing the diagnostics on the `SourceFile`, which can then be referenced here.